File: bench/regex.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# Regex - core Regexp (stdlib/language builtin), single realistic ops on a real doc
doc = File.read('../docs/learning-duso.md', encoding: 'UTF-8')

start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
has_datastore = !doc.match(/datastore\(/).nil?
contains_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start) * 1000

start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
words = doc.scan(/\w+/)
find_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start) * 1000

puts "contains() on #{doc.bytesize}-byte doc in #{contains_ms}ms (found=#{has_datastore})"
puts "find() on #{doc.bytesize}-byte doc in #{find_ms}ms (#{words.length} matches)"

File: bench/regex.js
// Regex - RegExp (stdlib/language builtin), single realistic ops on a real doc
const fs = require('fs');
const doc = fs.readFileSync('../docs/learning-duso.md', 'utf8');
const bytes = Buffer.byteLength(doc, 'utf8');

let start = process.hrtime.bigint();
const hasDatastore = /datastore\(/.test(doc);
const containsMs = Number(process.hrtime.bigint() - start) / 1e6;

start = process.hrtime.bigint();
const words = doc.match(/\w+/g) || [];
const findMs = Number(process.hrtime.bigint() - start) / 1e6;

console.log(`contains() on ${bytes}-byte doc in ${containsMs}ms (found=${hasDatastore})`);
console.log(`find() on ${bytes}-byte doc in ${findMs}ms (${words.length} matches)`);
